### PR TITLE
Final pass record layer cleanup

### PR DESCRIPTION
**Pull request by mattcaswell**
Following the record layer refactor this PR does a final pass cleanup on a number of items. The main changes are:

- Remove some fields from the SSL_CONNECTION structure which are no longer required
- Move the file tls_pad.c into ssl/record/methods - its only used from there
- Delete the ssl/record/ssl3_record.c file which is no longer needed
- Clean up declarations in ssl/record/record_local.h and ssl/record/record.h. Many declarations are no longer used, and many were better placed somewhere in ssl/record/methods
- Rename SSL3_BUFFER to TLS_BUFFER - the SSL3 prefix no longer seems approrpriate
- Rename SSL3_RECORD  to TLS_RL_RECORD - same reason as above. TLS_RL_RECORD was chosen over TLS_RECORD because the latter is already used elsewhere

This, combined with my other outstanding record layer related PRs (#19516 and #19550), represents the end of the record layer refactoring with the exception of one more PR which will update the design document to match the current state.

##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Comment by mattcaswell** [comment]
Fixed a memleak showing in the CIs.

File: ssl/record/methods/recmethod_local.h [hunk 1]
```diff
@@ -552,18 +552,18 @@ int tls_write_records_default(OSSL_RECORD_LAYER *rl,
...
+#define TLS_BUFFER_get_left(b)             ((b)->left)
+#define TLS_BUFFER_set_left(b, l)          ((b)->left = (l))
+#define TLS_BUFFER_sub_left(b, l)          ((b)->left -= (l))
+#define TLS_BUFFER_get_offset(b)           ((b)->offset)
+#define TLS_BUFFER_set_offset(b, o)        ((b)->offset = (o))
+#define TLS_BUFFER_add_offset(b, o)        ((b)->offset += (o))
+#define TLS_BUFFER_set_app_buffer(b, l)    ((b)->app_buffer = (l))
+#define TLS_BUFFER_is_app_buffer(b)        ((b)->app_buffer)
+
+void TLS_BUFFER_release(TLS_BUFFER *b);
```
Comment (t8m) [inline ssl/record/methods/recmethod_local.h:569]: Should it be named ossl_tls_buffer_release() instead?

File: ssl/record/methods/recmethod_local.h [hunk 2]
```diff
@@ -398,7 +398,7 @@ void ossl_rlayer_fatal(OSSL_RECORD_LAYER *rl, int al, int reason,
                                     || (rl)->version == TLS1_2_VERSION \
                                     || (rl)->isdtls)
 
-void SSL3_RECORD_set_seq_num(SSL3_RECORD *r, const unsigned char *seq_num);
+void TLS_RL_RECORD_set_seq_num(TLS_RL_RECORD *r, const unsigned char *seq_num);
```
Comment (t8m) [inline ssl/record/methods/recmethod_local.h:401]: ossl_tls_rl_record_set_seq_num()?

**Comment by mattcaswell** [comment]
Fixups pushed addressing feedback. @t8m please take another look. @hlandau - please reconfirm

**Review by t8m — APPROVED** [review]

**Review by hlandau — APPROVED** [review]
Still OK.

**Comment by hlandau** [comment]
Merged to master. Thank you.

**Pull request closed without merging**